### DISCUSSION TITLE
alevin: Question related to noisy reads, whitelisting, and emptyDrops

### DISCUSSION THREAD

**Discussion by imranfanaswala** [COMBINE-lab/salmon#506] [Q&A (single-cell specific)] [answered]
First of all, thank you for providing Salmon/Alevin and the related tools. 

Upon running Alevin (defaults for `--keepCBFraction` and `--expectCells`) on several samples; I noticed that:
1) For some samples, barely any cells were identified [1] (i.e. lots of noisy CBs); essentially rendering the entire sample useless.
2) For some samples, the number of CBs is greater than what we expect (i.e. they are not 5-10% of the cells that were "loaded").

One side note: I further filter the CBs for empty-droplets (using `emptyDrops()`). As I understand, eventhough whitelisting is unrelated to empty-droplets; `emptyDrops()` still assumes a high-proportion of low-confidence CBs in order to "work". And whitelisting collapses low-confidence CBs.

At this stage, here are my options:
a) Do nothing.  
b) Use Alevin's *final* whitelist on all samples; thereby retaining only high-confidence CBs. This hopefully resolves Issue (2). But if we do this, whats the risk (if any) of using `emptyDrops()` (recall that `emptyDrops()` expects low-confidence CBs too).
c) Do no whitelisting whatsoever (i.e. re-run Alevin with `--keepCBFraction 1` for all samples). Thereby, we rely on `emptyDrops()` entirely to identify the high-confidence CBs. Can't say if this will resolve either issues
d) Do (c) on only those samples with too many "noisy CBs" [1]. This hopefully resolves Issue 1
e) Do (b) on only those samples with higher-than-expected CBs. This hopefully resolves Issue 2

Somewhere in here there is also `--expectCells`, but I can't quite fit it in...

[1]
```
15.1609% reads will be thrown away because of noisy Cellular barcodes.
18.6041% reads will be thrown away because of noisy Cellular barcodes.
12.5753% reads will be thrown away because of noisy Cellular barcodes.
21.0916% reads will be thrown away because of noisy Cellular barcodes.
41.2315% reads will be thrown away because of noisy Cellular barcodes.
26.3949% reads will be thrown away because of noisy Cellular barcodes.
32.4048% reads will be thrown away because of noisy Cellular barcodes.
23.5219% reads will be thrown away because of noisy Cellular barcodes.
23.6427% reads will be thrown away because of noisy Cellular barcodes.
17.4213% reads will be thrown away because of noisy Cellular barcodes.
12.7759% reads will be thrown away because of noisy Cellular barcodes.
22.795% reads will be thrown away because of noisy Cellular barcodes.
14.6121% reads will be thrown away because of noisy Cellular barcodes.
97.3112% reads will be thrown away because of noisy Cellular barcodes.
26.8011% reads will be thrown away because of noisy Cellular barcodes.
17.8044% reads will be thrown away because of noisy Cellular barcodes.
31.952% reads will be thrown away because of noisy Cellular barcodes.
19.2951% reads will be thrown away because of noisy Cellular barcodes.
18.1361% reads will be thrown away because of noisy Cellular barcodes.
19.3543% reads will be thrown away because of noisy Cellular barcodes.
21.0769% reads will be thrown away because of noisy Cellular barcodes.
18.1759% reads will be thrown away because of noisy Cellular barcodes.
26.1144% reads will be thrown away because of noisy Cellular barcodes.
31.8096% reads will be thrown away because of noisy Cellular barcodes.
17.9811% reads will be thrown away because of noisy Cellular barcodes.
35.1091% reads will be thrown away because of noisy Cellular barcodes.
16.0757% reads will be thrown away because of noisy Cellular barcodes.
31.0397% reads will be thrown away because of noisy Cellular barcodes.
22.4944% reads will be thrown away because of noisy Cellular barcodes.
27.2131% reads will be thrown away because of noisy Cellular barcodes.
11.8507% reads will be thrown away because of noisy Cellular barcodes.
12.7244% reads will be thrown away because of noisy Cellular barcodes.
12.9291% reads will be thrown away because of noisy Cellular barcodes.
24.9483% reads will be thrown away because of noisy Cellular barcodes.
19.9322% reads will be thrown away because of noisy Cellular barcodes.
48.2229% reads will be thrown away because of noisy Cellular barcodes.
35.8745% reads will be thrown away because of noisy Cellular barcodes.
21.8517% reads will be thrown away because of noisy Cellular barcodes.
26.8803% reads will be thrown away because of noisy Cellular barcodes.
42.3376% reads will be thrown away because of noisy Cellular barcodes.
21.1344% reads will be thrown away because of noisy Cellular barcodes.
23.3274% reads will be thrown away because of noisy Cellular barcodes.
22.3853% reads will be thrown away because of noisy Cellular barcodes.
16.0405% reads will be thrown away because of noisy Cellular barcodes.
21.141% reads will be thrown away because of noisy Cellular barcodes.
14.3761% reads will be thrown away because of noisy Cellular barcodes.
49.9611% reads will be thrown away because of noisy Cellular barcodes.
33.1622% reads will be thrown away because of noisy Cellular barcodes.
18.2677% reads will be thrown away because of noisy Cellular barcodes.
22.8454% reads will be thrown away because of noisy Cellular barcodes.
22.2167% reads will be thrown away because of noisy Cellular barcodes.
12.1184% reads will be thrown away because of noisy Cellular barcodes.
13.3562% reads will be thrown away because of noisy Cellular barcodes.
10.4225% reads will be thrown away because of noisy Cellular barcodes.
21.2565% reads will be thrown away because of noisy Cellular barcodes.
16.3058% reads will be thrown away because of noisy Cellular barcodes.
17.609% reads will be thrown away because of noisy Cellular barcodes.
30.9462% reads will be thrown away because of noisy Cellular barcodes.
47.0413% reads will be thrown away because of noisy Cellular barcodes.
11.6665% reads will be thrown away because of noisy Cellular barcodes.
26.6606% reads will be thrown away because of noisy Cellular barcodes.
```

**Comment by k3yavi** [comment, marked as answer]
Hi @imranfanaswala ,

Thanks for raising the important question. The short answer is that the defaults are designed with a generic case in mind. While they may works reasonably well for most cases, it's hard to come up with the parameters that work well for all cases and we might have to tweak a few things.

The question, I think, is why do we see such a variation in the number of quantified cells (too high or low) and the answer is there can be multiple reasons. The first thing which can be off is the initial whitelisting. Alevin performs a _knee_ based thresholding in the cumulative frequency distribution of the observed cellular barcodes. In a relatively clean dataset, it's easy to find the cutoff but it's a heuristic, which can over and under shoot sometimes. The first thing to monitor is `X% reads will be thrown away because of noisy Cellular barcodes`. If the number is too high (generally stays below 20%) then initial whitelisting is overshooting and alevin needs help figuring our the number of expected cells where the flag `--expectCells` comes into scene. 

The `--expectCells Y` follows the CellRanger type method of taking top 10% cells and using the 90% cbs within the frequency range of the provided threshold of `Y`. The option here is to:
1.) selectively apply the flag for cases where initial whitelisting is over shooting. 
2.) always provide the `--expectCells` option 
3.) Use `--forceCells Z`: this forces alevin to quantify top `Z` cells.

Now the options relies on user's choice and the data being quantified. If you wan't completely automated pipeline with no user intervention at all, then I think option c (in your text) will solve both your issues. However, if you can monitor some of the stats then I propose:
1.) Run regular alevin, if that fails due to too many or too low cbs then
2.) Provide `--expectCells Y`, where Y would be the number of expected cells, if this fails then,
3.) Provide `--forceCells Z`.

Hope it helps.

**Comment by cnk113** [comment]
I had a few questions on running alevin on 10X V3, should I still be using ISR for lib type and when I set forceCells does it take the top cells by UMI count?

**Comment by k3yavi** [comment]
Hi @cnk113 ,
Yes, to my best knowledge, 10 v3 is also a stranded protocol and libtype should be set as ISR, with --chromiumV3 flag 

ForceCells is applied on the number of observed *reads* for each cellular barcode.
